subsys/net/lib/nrf_cloud/src/nrf_cloud_log_backend.c

```diff
@@ -381,69 +445,96 @@ static int send_ring_buffer(void)
 	}
 
 	stored = ring_buf_size_get(&log_nrf_cloud_rb);
-	output.data.len = ring_buf_get_claim(&log_nrf_cloud_rb,
-					     (uint8_t **)&output.data.ptr,
-					     stored);
-	if (output.data.len != stored) {
+	log_rb_len = ring_buf_get_claim(&log_nrf_cloud_rb, &log_rb_ptr, stored);
+	if (log_rb_len != stored) {
 		LOG_WRN("Capacity:%u, free:%u, stored:%u, claimed:%u",
 			ring_buf_capacity_get(&log_nrf_cloud_rb),
 			ring_buf_space_get(&log_nrf_cloud_rb),
-			stored, output.data.len);
-		stored = output.data.len;
+			stored, log_rb_len);
+		stored = log_rb_len;
 	}
-	if (!output.data.len) {
+	if (!log_rb_len) {
 		goto cleanup;
 	}
+	if ((log_format_current == LOG_OUTPUT_DICT) &&
+	    IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		topic = get_dict_topic();
+		if (!topic) {
+			goto cleanup;
+		}
+		err = convert_to_quoted_base64(log_rb_ptr, log_rb_len,
+					       &log_b64_ptr, &log_b64_len);
+		if (err) {
+			goto cleanup;
+		}
+		LOG_DBG("topic:   %s", topic);
+		LOG_DBG("payload: %s", (const char *)log_b64_ptr);
+		output_data.ptr = log_b64_ptr;
+		output_data.len = log_b64_len;
+	} else {
+		/* Send ring buffer contents as is -- JSON or raw binary. */
+		log_rb_ptr[log_rb_len] = '\0';
+		output_data.ptr = log_rb_ptr;
+		output_data.len = log_rb_len;
+	}
 
-	uint8_t *p = (uint8_t *)output.data.ptr;
-
-	p[output.data.len] = '\0';
-
-	LOG_DBG("Ready to transmit %zd bytes...", output.data.len);
+	LOG_DBG("Ready to transmit %zd bytes...", output_data.len);
 	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
+		struct nrf_cloud_tx_data output = {
+			.qos = MQTT_QOS_0_AT_MOST_ONCE,
+			.topic_type = (log_format_current == LOG_OUTPUT_TEXT) ?
+				       NRF_CLOUD_TOPIC_BULK : NRF_CLOUD_TOPIC_BIN,
+			.data.ptr = output_data.ptr,
+			.data.len = output_data.len
+		};
+
 		err = nrf_cloud_send(&output);
 	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
 		do {
+			/* If no topic provided, use d2c/bulk, otherwise use topic. */
 			err = nrf_cloud_rest_send_device_message(log_context.rest_ctx,
 								 log_context.device_id,
-								 output.data.ptr, true, NULL);
+								 output_data.ptr,
+								 !topic, topic);
 			if (err) {
 				LOG_ERR("Error sending message:%d", err);
-				if (p[0] == '\0') {
-					LOG_ERR("Empty buffer!");
-				}
 				if (err != -EBUSY) {
 					LOG_ERR("Data: %s, len: %zd",
-						(const char *)output.data.ptr, output.data.len);
+						(const char *)output_data.ptr, output_data.len);
 				}
 				k_sleep(K_MSEC(100));
 			}
 		} while (err == -EBUSY);
+		if (log_b64_ptr) {
+			nrf_cloud_free(log_b64_ptr);
+			log_b64_ptr = NULL;
+		}
```

Comment:
Possibly move this to after the `cleanup` label, since it can live there with no problem